### PR TITLE
Supporting Intel AMX instructions in quantized GEMM

### PR DESCRIPTION
**Pull request by chenfucn**
### Description
Using Intel AMX int8 instructions to accelerate quantized GEMM

### Motivation and Context
AMX instructions accelerate quantized GEMM significantly:

Prepacked B perf numbers (latency in ns)

GEMM Config | AVX512Vnni | AMX
-- | --: | --:
M:384/N:1024/K:1024/Batch:1/Threads:4 | 1057511 | 285393
M:384/N:1024/K:3072/Batch:1/Threads:4 | 2643929 | 700397
M:384/N:1024/K:4096/Batch:1/Threads:4 | 3784750 | 890701
M:384/N:4096/K:1024/Batch:1/Threads:4 | 2378139 | 887251
M:384/N:1024/K:1024/Batch:1/Threads:16 | 307137 | 138481
M:384/N:1024/K:3072/Batch:1/Threads:16 | 855730 | 295027
M:384/N:1024/K:4096/Batch:1/Threads:16 | 1126878 | 317395
M:384/N:4096/K:1024/Batch:1/Threads:16 | 781963 | 237014
M:1536/N:1024/K:1024/Batch:1/Threads:16 | 538864 | 181459
M:1536/N:1024/K:3072/Batch:1/Threads:16 | 1681002 | 561600
M:1536/N:1024/K:4096/Batch:1/Threads:16 | 2158127 | 717470
M:1536/N:4096/K:1024/Batch:1/Threads:16 | 2428622 | 896140
M:3072/N:1024/K:1024/Batch:1/Threads:16 | 1058029 | 357031
M:3072/N:1024/K:3072/Batch:1/Threads:16 | 3138504 | 1095857
M:3072/N:1024/K:4096/Batch:1/Threads:16 | 4155640 | 1386183
M:3072/N:4096/K:1024/Batch:1/Threads:16 | 4679030 | 1778624

### CODE REVIEW

**Review by jchen351 — APPROVED** [review]

File: onnxruntime/core/mlas/lib/qgemm_kernel_amx.cpp [hunk 1]
```diff
@@ -0,0 +1,730 @@
...
+
+    static constexpr size_t PackedK = TILE_K;
+
+    // Use smaller stride for debugging,
+    static constexpr MLAS_GEMM_QUANT_STRIDES Strides{32, 128, 1024};
+    static constexpr MLAS_GEMM_QUANT_STRIDES PackedStrides{32, 512, 2048};
+};
+
+constexpr size_t MLAS_GEMM_U8S8_KERNEL_AMX::PackedK;
+constexpr MLAS_GEMM_QUANT_STRIDES MLAS_GEMM_U8S8_KERNEL_AMX::Strides;
```
Comment (snnn) [inline onnxruntime/core/mlas/lib/qgemm_kernel_amx.cpp:53]: You may try this: https://www.codingame.com/playgrounds/2205/7-features-of-c17-that-will-simplify-your-code/inline-variables

Comment (chenfucn) [reply onnxruntime/core/mlas/lib/qgemm_kernel_amx.cpp:53]: hmm, this pattern happens a lot in old MLAS. let me get another PR for this.

**Review by jchen351 — APPROVED** [review]

File: cmake/onnxruntime_mlas.cmake [hunk 2]
```diff
@@ -3,6 +3,18 @@
 
 set(MLAS_SRC_DIR ${ONNXRUNTIME_ROOT}/core/mlas/lib)
 
+
+set(MLAS_AMX_SUPPORTED FALSE)
+
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 11)
+  set(MLAS_AMX_SUPPORTED TRUE)
```
Comment (yufenglee) [inline cmake/onnxruntime_mlas.cmake:10]: >set(MLAS_AMX_SUPPORTED TRUE) [](http://example.com/codeflow?start=2&length=28)

print a warning if the compiler doesn't support AMX

File: onnxruntime/core/mlas/lib/qgemm_kernel_amx.cpp [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+ * @brief move data from Tile buffer to C
+ *
+ */
+static inline
+void
+MoveTile(const int32_t* Tile, size_t cntM, uint16_t MaskN, int32_t* c_ptr, size_t ldc)
+{
+    for (size_t i = 0; i < cntM; i++){
+        __m512i c = _mm512_maskz_loadu_epi32(MaskN, Tile);
+        Tile += 16;
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/qgemm_kernel_amx.cpp:378]: >16 [](http://example.com/codeflow?start=16&length=2)

nit: TILE_N

File: onnxruntime/core/mlas/lib/amd64/QgemmU8S8KernelAmx.asm [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+        vpcmpeqw ymm0,ymm0,ymm0             ; generate word vector [0xFFFF]
+        vpsrlw  ymm0,ymm0,15                ; generate word vector [0x0001]
+        vpsllw  ymm1,ymm0,8                 ; generate word vector [0x0100]
+        vpor    ymm1,ymm0,ymm1              ; generate word vector [0x0101]
+
+;
+; Compute the bit flip vector to adjust input from U8 to S8.
+;
+
+        vpxor   xmm2,xmm2,xmm2              ; generate word vector [0x0000]
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/amd64/QgemmU8S8KernelAmx.asm:103]: >vpxor   xmm2,xmm2,xmm2              ; generate word vector [0x0000] [](http://example.com/codeflow?start=8&length=67)

nit: there's no need to maintain a xmm2 with 0. instruction below 'vmovaps xmm17,xmm2' can use vpxor to clear the register.

Comment (chenfucn) [reply onnxruntime/core/mlas/lib/amd64/QgemmU8S8KernelAmx.asm:103]: it's actually clearing out ymm2, which is either 0x0000 or 0x0808 based on whether B is signed or not. This is used for optional 8b top bit flip in line 144 and 145

Comment (yufenglee) [reply onnxruntime/core/mlas/lib/amd64/QgemmU8S8KernelAmx.asm:103]: I see. Missed it.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**